tests/describe_task_queue_test.go

```diff
@@ -229,14 +242,49 @@ func (s *DescribeTaskQueueSuite) validateDescribeTaskQueue(
 	expectedAddRate map[enumspb.TaskQueueType]bool,
 	expectedDispatchRate map[enumspb.TaskQueueType]bool,
 	isEnhancedMode bool,
+	isCached bool,
 ) {
 	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
 	defer cancel()
 
 	var resp *workflowservice.DescribeTaskQueueResponse
 	var err error
 
-	if isEnhancedMode {
+	if isCached {
+		resp, err = s.FrontendClient().DescribeTaskQueue(ctx, &workflowservice.DescribeTaskQueueRequest{
+			Namespace:              s.Namespace(),
+			TaskQueue:              &taskqueuepb.TaskQueue{Name: tq, Kind: enumspb.TASK_QUEUE_KIND_NORMAL},
+			ApiMode:                enumspb.DESCRIBE_TASK_QUEUE_MODE_ENHANCED,
+			Versions:               nil, // default version, in this case unversioned queue
+			TaskQueueTypes:         nil, // both types
+			ReportPollers:          true,
+			ReportTaskReachability: false,
+			ReportStats:            true,
+		})
+		s.NoError(err)
+		s.NotNil(resp)
+		s.Equal(1, len(resp.GetVersionsInfo()), "should be 1 because only default/unversioned queue")
+		versionInfo := resp.GetVersionsInfo()[""]
+		s.Equal(enumspb.BUILD_ID_TASK_REACHABILITY_UNSPECIFIED, versionInfo.GetTaskReachability())
+		types := versionInfo.GetTypesInfo()
+		s.Equal(len(types), len(expectedBacklogCount))
+
+		wfStats := types[int32(enumspb.TASK_QUEUE_TYPE_WORKFLOW)].Stats
+		actStats := types[int32(enumspb.TASK_QUEUE_TYPE_ACTIVITY)].Stats
+
+		// Actual counter can be greater than the expected due to history retries. We make sure the counter is in
+		// range [expected, expected+maxBacklogExtraTasks]
+		s.GreaterOrEqual(wfStats.ApproximateBacklogCount, expectedBacklogCount[enumspb.TASK_QUEUE_TYPE_WORKFLOW])
+		s.LessOrEqual(wfStats.ApproximateBacklogCount, expectedBacklogCount[enumspb.TASK_QUEUE_TYPE_WORKFLOW]+maxBacklogExtraTasks[enumspb.TASK_QUEUE_TYPE_WORKFLOW])
+		s.GreaterOrEqual(actStats.ApproximateBacklogCount, expectedBacklogCount[enumspb.TASK_QUEUE_TYPE_ACTIVITY])
+		s.LessOrEqual(actStats.ApproximateBacklogCount, expectedBacklogCount[enumspb.TASK_QUEUE_TYPE_ACTIVITY]+maxBacklogExtraTasks[enumspb.TASK_QUEUE_TYPE_ACTIVITY])
+		s.Equal(wfStats.ApproximateBacklogCount == 0, wfStats.ApproximateBacklogAge.AsDuration() == time.Duration(0))
+		s.Equal(actStats.ApproximateBacklogCount == 0, actStats.ApproximateBacklogAge.AsDuration() == time.Duration(0))
+		s.Equal(expectedAddRate[enumspb.TASK_QUEUE_TYPE_WORKFLOW], wfStats.TasksAddRate > 0)
+		s.Equal(expectedAddRate[enumspb.TASK_QUEUE_TYPE_ACTIVITY], actStats.TasksAddRate > 0)
+		s.Equal(expectedDispatchRate[enumspb.TASK_QUEUE_TYPE_WORKFLOW], wfStats.TasksDispatchRate > 0)
+		s.Equal(expectedDispatchRate[enumspb.TASK_QUEUE_TYPE_ACTIVITY], actStats.TasksDispatchRate > 0)
+	} else if isEnhancedMode {
```

Comment:
should this be an `else if` or just a separate `if`? aren't there many test scenarios with both caching and enhanced mode?

Reply:
not currently - right now, tests testing the caching behaviour call `DescribeTaskQueue` without any retries. This has been done because we load task-queue statistics when we call `DescribeTaskQueuePartition` itself (which has retries). Thus, we need an `else if` since code inside the else consists of describe task queue calls with retries.

A potential question could be - Why do we load task-queue statistics by calling `DescribeTaskQueuePartition` before `DescribeTaskQueue` for these caching tests. This was done after having discussions with @ShahabT and was because:

1. The caching behaviour is implemented inside DescribeTaskQueue and we wanted to directly cross-verify if stats are fetched from the partition or fetched from the cache.